File: app/src/main/java/com/kaiserpudding/howtheywrite/characterDetail/CharacterDetailViewModelFactory.java
package com.kaiserpudding.howtheywrite.characterDetail;

import android.app.Application;
import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.support.annotation.NonNull;

public class CharacterDetailViewModelFactory implements ViewModelProvider.Factory {

  private Application application;
  private int characterId;

  public CharacterDetailViewModelFactory(Application application, int characterId) {
    this.application = application;
    this.characterId = characterId;
  }

  @NonNull
  @Override
  public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
    return (T) new CharacterDetailViewModel(application, characterId);
  }
}

File: HowTheyWrite_android/app/src/main/java/com/kaiserpudding/howtheywrite/characterDetail/CharacterDetailViewModelFactory.kt
package com.kaiserpudding.howtheywrite.characterDetail

import android.app.Application
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider

/**
 * Factory to create a [CharacterDetailViewModel] with an additional parameter
 *
 * @property application
 * @property characterId
 */
@Suppress("UNCHECKED_CAST")
class CharacterDetailViewModelFactory(private val application: Application, private val characterId: Int) : ViewModelProvider.Factory {

    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        return CharacterDetailViewModel(application, characterId) as T
    }
}
